Added a task to unphase genotypes before HiPhase.

hangsuUNC opened the pull request:
Add a task to use bcftools to clean phasing information before hiphase.

samuelklee commented:
Thanks for adding this! I have some comments but am currently making my way back after some travel delays, so please don’t merge yet.

Have you evaluated the effect of the change yet using PhasedPanelEvaluation? If so, please add a summary of the relevant metrics here.

Also, please rename the PR and revise the commit messages according to that WARP style guide linked elsewhere. Take a look at my PRs/commits to get an idea.

samuelklee commented:
Just saw your last run. Don’t forget to explore the effect of Shapeit4 reproducibility. You may want to rerun the baseline single-threaded.

samuelklee commented:
Kicked off some runs that (hopefully) account for Shapeit4 reproducibility here:

main w/ num_t = 1: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/c32db516-895e-4b25-bc95-0faac54d0728
main w/ num_t = 1, no call caching (to test reproducibility): https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/2dc74da8-56b1-45d3-b364-6e0788972d82
this branch w/ num_t = 1: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/b969719e-e485-410d-bc35-b332ec63293e

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -47,6 +47,13 @@ workflow PhysicalAndStatisticalPhasing {
         locus = region
     }
 
+    call FilterPhasingInfo as FilterPhasing { input:
```
I would not call this operation "Filtering" as that implies hard/soft removal of entire records, whereas here we are just modifying genotypes. "Info" is also (ironically!) noninformative. `UnphaseGenotypes` would be more clear and succinct, IMO, but if you come up with something better, feel free to go with that!

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -47,6 +47,13 @@ workflow PhysicalAndStatisticalPhasing {
+        sv_vcf = SubsetVcfSV.subset_vcf,
+        sv_vcf_tbi = SubsetVcfSV.subset_tbi,
+        prefix = prefix + "filter_phasing"
+
```
Extra white space.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
 }
+
+
+# filter out singletons (i.e., keep MAC >= 2) and concatenate with deduplication
```
Vestigial comment.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+    command <<<
+        set -euxo pipefail
+
+        # filter SV singletons
```
Vestigial comment.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+        bootDiskSizeGb: 10
+        preemptible_tries:     3
+        max_retries:           2
+        docker:"hangsuunc/cleanvcf:v1"
```
We should avoid personal Docker images in production code. This task simply uses bcftools as does `FilterAndConcatVcfs`, so is there no reason they cannot use the same Docker (e.g., different bcftools versions)?

hangsuUNC replied:
I tried "us.gcr.io/broad-dsp-lrma/lr-gcloud-samtools:0.1.2", it seems bcftools +setGT doesn't work for this version of bcftools. Should we update the public docker somehow?

samuelklee replied:
I don't think it is a version issue, since that Docker contains bcftools 1.18 and the +setGT plugin predates that. Rather, I think that bcftools in the Docker might not be set up to use *any* plugins. @SHuang-Broad might have some opinions on what to do here.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+    >>>
+
+    output {
+        File filter_phasing_vcf = "~{prefix}.vcf.gz"
```
Should be`unphased_sv_vcf` and `unphased_sv_vcf_tbi`.

samuelklee replied:
(Or more generically 'unphased_vcf`, etc.---see comment elsewhere.)

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -47,6 +47,13 @@ workflow PhysicalAndStatisticalPhasing {
+    call FilterPhasingInfo as FilterPhasing { input:
+        sv_vcf = SubsetVcfSV.subset_vcf,
+        sv_vcf_tbi = SubsetVcfSV.subset_tbi,
+        prefix = prefix + "filter_phasing"
```
This does not respect the existing prefix convention. Since this will only be applied to the SV VCF, I think `prefix + ".unphased"` is fine.

samuelklee replied:
Note also that I neglected to change the prefix convention in ConvertLowerCase. I think `~{samplename}.uppercased` would be most succinct there.

Note also that some tasks enforce the prefix convention in their output lines (as does ConvertLowerCase), while others expose them and enforce the convention in the call to the task. Perhaps we should do this more consistently.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+
+        # filter SV singletons
+        bcftools +setGT ~{sv_vcf} -- -t a -n u | bgzip > ~{prefix}.vcf.gz
+        bcftools index --tbi ~{prefix}.vcf.gz
```
`bcftools index -t` is used elsewhere.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+        # filter SV singletons
+        bcftools +setGT ~{sv_vcf} -- -t a -n u | bgzip > ~{prefix}.vcf.gz
+        bcftools index --tbi ~{prefix}.vcf.gz
+
```
White space.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+task FilterPhasingInfo {
+
+    input {
+        File sv_vcf            # biallelic
```
The `biallelic` comment is also vestigial.

If we want this to be a generic `UnphaseGenotypes` task that doesn't care if a VCF contains short/SV bi/multiallelic records, we should name the inputs in a generic fashion accordingly. If so, perhaps when we call the task we can alias it as `UnphaseSVGenotypes` to hint at the fact that we don't apply this operation to the short variants.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +232,39 @@ task FilterAndConcatVcfs {
+        set -euxo pipefail
+
+        # filter SV singletons
+        bcftools +setGT ~{sv_vcf} -- -t a -n u | bgzip > ~{prefix}.vcf.gz
```
A comment would be useful to document the meaning of `-t a -n u`. Also `-Oz -o` is used elsewhere, rather than piping to `bgzip`. Please also format the command line and any line breaks as elsewhere.

samuelklee reviewed (changes requested):
Some comments, mostly all style related. Let's also record here the results of the tests I just kicked off, when they're finished. Thanks again!

hangsuUNC commented:
Some test results:

Before clean phased genotypes, the HiphaseSV phasing_summary:
PHASE_BLOCKS	SWITCH_ERRORS	FLIP_ERRORS	NG_50	SWITCH_NGC50	SWITCHFLIP_NGC50
6	0	1	0	0	0

After clean phased genotypes, the Hiphase SV phasing_summary:
PHASE_BLOCKS	SWITCH_ERRORS	FLIP_ERRORS	NG_50	SWITCH_NGC50	SWITCHFLIP_NGC50
6	0	0	0	0	0

In summary, one flip error is corrected by filtering out the pre-hiphase phased genotypes.

Let me know if you have any comments!

Hang

samuelklee commented:
Thanks for recording those. I assume they are from the original `num_t = 8` main vs. branch runs, since the `num_t = 1` runs I kicked off have not finished yet? In the future, it might be nice to link explicitly to the original submissions or files (although note that, in general, files like the latter might get blown away, e.g. during workspace cleanups).

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -71,15 +77,15 @@ workflow PhysicalAndStatisticalPhasing {
             input:
                 vcf = SplitVcfbySampleSV.single_sample_vcf,
-                samplename = sample_id
+                samplename = sample_id + ".uppercased"
```
Right, looking at everything again, perhaps `samplename` -> `prefix` here and below and remove all `_sv_cleaned` below.

samuelklee commented:
Looks like all test runs have completed. From looking across the 4 runs, I am not sure we can definitively say whether this change is an improvement? I see a drop in recall after the HiPhaseSV stage, and identical PR after the Shapeit4 stage. Do you have an idea about how this drop should be weighted against the changes to switch/flip errors?

In any case, we should certainly run the tool in the manner suggested by the authors. So we can merge this after remaining comments are addressed. But all of this highlights that HiPhase is perhaps partially redundant with the upstream pipeline.

samuelklee commented on wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl:
```diff
@@ -225,3 +231,36 @@ task FilterAndConcatVcfs {
         docker:"us.gcr.io/broad-dsp-lrma/lr-gcloud-samtools:0.1.2"
     }
 }
+
```
White space.

samuelklee commented:
@hangsuUNC looks like there is still an unaddressed comment. I also see one or two more things that I can clean up. Would you mind if I address and get this in later today? Thanks again for putting it together!

hangsuUNC commented:
> @hangsuUNC looks like there is still an unaddressed comment. I also see one or two more things that I can clean up. Would you mind if I address and get this in later today? Thanks again for putting it together!

Of course! Sorry if there's something I'm missing. Please feel free to make edits!

Thanks,

Hang

samuelklee commented:
Rebased and added a few more minor style fixes. Will run PhasedPanelEvaluation, to be compared against this baseline: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/135f1d17-5214-43c2-8cfa-9716d3784ca3

samuelklee commented:
Comparing e.g. EvaluateHiPhaseSV results for [main](gs://fc-c6ae4a57-f44e-4d6c-befd-c48a2abcecc7/submissions/135f1d17-5214-43c2-8cfa-9716d3784ca3/PhasedPanelEvaluation/10e7b307-35fc-4147-bcdf-70e36ac5d0e8/call-EvaluateHiPhaseSV/VcfdistAndOverlapMetricsEvaluation/6d485d3f-d946-4d53-9854-4928fa9bbdab/call-Vcfdist/shard-0/cacheCopy) vs. [this branch](gs://fc-c6ae4a57-f44e-4d6c-befd-c48a2abcecc7/submissions/ba390f1d-82f2-403b-a00b-4c3fe795a5e3/PhasedPanelEvaluation/0fddbe5c-83fc-4ac7-b978-19eb0a7121f6/call-EvaluateHiPhaseSV/VcfdistAndOverlapMetricsEvaluation/5ec529de-22dc-4939-b192-e655f3fb330c/call-Vcfdist/shard-0), we see mixed results. For example, the number of flip errors is reduced from 1 to 0, but the SV recall also drops from 71% to 61%. The number of inconsistent sites/alleles stays the same.

But again, if the authors of the tool say that inputs are expected to be unphased, let's respect that. We should perhaps evaluate on a larger region or after we switch to kanpig intra+inter later on.
